buildpacks/ruby/src/layers/ruby_install_layer.rs

```diff
@@ -362,4 +322,33 @@ version = "3.1.3"
             "https://heroku-buildpack-ruby.s3.us-east-1.amazonaws.com/heroku-22/ruby-2.7.4.tgz",
         );
     }
+
+    #[test]
+    fn test_ruby_version_difference_clears_cache() {
+        let old = Metadata {
+            ruby_version: ResolvedRubyVersion("2.7.2".to_string()),
+            distro_name: "ubuntu".to_string(),
+            distro_version: "20.04".to_string(),
+            cpu_architecture: "x86_64".to_string(),
+        };
+        let differences = old.diff(&old);
```

Comment:
I think this test combines two distinct things that should be validated separately:
- The `diff` behavior should have it's own set of tests for the `MetadataDiff` trait implementation
- The `restored_layer_action` should have it's own set of of tests for cache layer action + expected output format

Reply:
> The restored_layer_action should have it's own set of of tests for cache layer action + expected output format

I think that's a good idea. I'm trying to balance showing "just enough" of the functionality with a specific use case and was worried that a ton of tests make the PR line numbers seem larger than they really are. Via your comment it sounds like you're onboard with the approach and think it's worth it to invest more in stabilization and guarantees. If so, I agree.

Reply:
> The diff behavior should have it's own set of tests for the MetadataDiff trait implementation

I'm not totally following here. Are you saying we should also test `MetadataDiff` in the shared module? If so: to test `restored_layer_action` in isolation I'll need to exercise the MetadataDiff trait. I think what you're wanting will fall out naturally, but the logic in those tests will be inherently coupled for now (which is okay, because they're two concepts that are coupled).

An alternative read could be that you feel the ruby layer module shouldn't have knowledge of the `restored_layer_action`.

If you meant something else, let me know.

Reply:
> An alternative read is that you're wanting to not call restored_layer_action from these tests and instead only test the MetadataDiff logic here and trust that restored_layer_action will do the right thing which is what I started out with.

That one. I don't mind seeing more unit test code. It's nice to see how various permutations and edge cases play out. 

Also, I like how you've structured the code so that it's possible to isolate out these relevant parts 😄

Reply:
That makes sense. I view vaguely feel that testing only MetadataDiff as relying too much on the implementation. If change that Trait in some way that doesn't modify `diff` behavior but makes `restored_layer_action` not rely on it, then we're not asserting the end behavior we want. Unfortunately that desire to "test the behavior and not the implementation" is a bit of a rabbit hole.

I spent, perhaps too much time thinking on how to best structure this so that we rely on minimal coupling. It's a bit verbose so I put it in it's own gist https://gist.github.com/schneems/a4b7074793e8fdc1074c16f880c178f7 but if you want to chat about it, let's do that here. Interestingly enough that led me back to one of my first closed PRs against libcnb which was the desire to include some kind of `BuildContext` test generator.

Reply:
I introduced a shorthand way to make a BuildContext that lets me exercise the MetadataDiff logic. https://github.com/heroku/buildpacks-ruby/pull/327/commits/36d14eb820ee2c1d6207173bb1c0d23fc0cfb3f4

I'll need to use another approach for the invalid metadata. Working on that now.

Reply:
Ready for re-review. I added tests to shared and changed the interface I'm testing in the `ruby_install_layer`. 

The PR size ballooned, but 209 of the 397 lines are tests and helpers.

Please take a look. Also, @runesoerensen, this is ready for review if you can.

Reply:
To bring it back, i ended up implementing the ability to call the top level function that's also being used in the layer which requires a BuildContext. Which is overkill, but I needed it anyway to be able to write tests for the shared logic.

I also added (just now) tests for the format of the MetadataDiff. I think in the future, we can rely on that interface for testing and leave a comment mentioning that the logic doubles up as cache invalidation logic.

Thanks for the review and suggestions!